hawkbit-repository/hawkbit-repository-jpa/src/test/java/org/eclipse/hawkbit/repository/jpa/ControllerManagementTest.java

```diff
@@ -497,19 +506,111 @@ public void hasTargetArtifactAssignedIsTrueWithMultipleArtifacts() {
+            fail("Expected an ConcurrencyFailureException to be thrown!");
+        } catch (ConcurrencyFailureException e) {
+            verify(mockTargetRepository, times(TX_RT_MAX)).findOne(any());
+        } finally {
+            // revert
+            ((JpaControllerManagement) controllerManagement).setTargetRepository(targetRepository);
+        }
+    }
+
+    @Test
+    @Description("Register a controller which does not exist, when a ConcurrencyFailureException is raised, the " +
+            "exception is not rethrown when the max retries are not yet reached")
+    @ExpectEvents({ @Expect(type = TargetCreatedEvent.class, count = 1),
+            @Expect(type = TargetPollEvent.class, count = 1) })
+    public void findOrRegisterTargetIfItDoesNotExistDoesNotThrowExceptionBeforeMaxRetries() {
+
+        TargetRepository mockTargetRepository = Mockito.mock(TargetRepository.class);
+        ((JpaControllerManagement) controllerManagement).setTargetRepository(mockTargetRepository);
+        final Target target = testdataFactory.createTarget();
+
+        when(mockTargetRepository.findOne(any()))
+                .thenThrow(new ConcurrencyFailureException("oops!"))
+                .thenThrow(new ConcurrencyFailureException("oops!"))
+                .thenReturn(Optional.of((JpaTarget) target));
+        when(mockTargetRepository.save(any())).thenReturn(target);
+
+
+        final Target targetFromControllerManagement = controllerManagement
+                .findOrRegisterTargetIfItDoesNotExist(target.getControllerId(), LOCALHOST);
+
+        verify(mockTargetRepository, times(3)).findOne(any());
+        verify(mockTargetRepository, times(1)).save(any());
+        assertEquals(target, targetFromControllerManagement);
+
+        // revert
+        ((JpaControllerManagement) controllerManagement).setTargetRepository(targetRepository);
+    }
+
+    @Test
+    @Description("Register a controller which does not exist, if a EntityAlreadyExistsException is raised, the " +
+            "exception is rethrown and no further retries will be attempted")
+    public void findOrRegisterTargetIfItDoesNotExistDoesntRetryWhenEntityAlreadyExistsException() {
+
+        TargetRepository mockTargetRepository = Mockito.mock(TargetRepository.class);
+        ((JpaControllerManagement) controllerManagement).setTargetRepository(mockTargetRepository);
+
+        when(mockTargetRepository.findOne(any())).thenReturn(Optional.empty());
+        when(mockTargetRepository.save(any())).thenThrow(new EntityAlreadyExistsException());
+
+        try {
+            controllerManagement.findOrRegisterTargetIfItDoesNotExist("1234", LOCALHOST);
+            fail("Expected an EntityAlreadyExistsException to be thrown!");
+        } catch (EntityAlreadyExistsException e) {
+            verify(mockTargetRepository, times(1)).findOne(any());
+            verify(mockTargetRepository, times(1)).save(any());
```

Comment:
actually not, because the event is created after a successful target creation, and here the mock will throw an `EntityAlreadyExists` exception once the `save()` method is called.